### PR TITLE
Implement generic scanning in Atomic

### PR DESCRIPTION
**Pull request by baude**
As more scanners besides openscap become available, atomic
can now begin to leverage them.  The new scan function has
been broken out into its on file (scan.py).

The scan command itself now defaults to openscap but can
also be switched to blackduck with --scanner.

Atomic now can use a configuration file which is stored
in /etc/atomic.conf.  In the case of the scan function,
we need the scanner defined in the configuration file
as well as the fully qualified image name and the
scan arguments.  Optionally, you can provide additional
custom docker arguments for the scanner as well

### CODE REVIEW

**Comment by baude** [comment]
I need to verify this with python3 yet but review can begin.

File: Atomic/atomic.py [hunk 1]
```diff
@@ -814,31 +721,40 @@ def get_image_infos(self):
...
+    def mount(self, mountpoint=None, docker_object=None, live=False):
         try:
-            options = [opt for opt in self.args.options.split(',') if opt]
-            mount.DockerMount(self.args.mountpoint,
-                              self.args.live).mount(self.args.image, options)
+            if mountpoint is None:
+                mountpoint = self.args.mountpoint
+            if docker_object is None:
+                docker_object = self.args.image
+            if 'live' in self.args:
```
Comment (rhatdan) [inline Atomic/atomic.py]: Why not call this image?

File: Atomic/atomic.py [hunk 2]
```diff
@@ -814,31 +721,40 @@ def get_image_infos(self):
...
+                live = self.args.live
+
+            options = [opt for opt in self.args.options.split(',') if opt] if 'options' in self.args else ""
+            mount.DockerMount(mountpoint, live).mount(docker_object,
+                                                                options)
```
Comment (rhatdan) [inline Atomic/atomic.py]: Put on same line.

File: Atomic/util.py [hunk 3]
```diff
@@ -213,6 +118,7 @@ def get_mounts_by_path():
                           )
     return mount_info
 
+
```
Comment (rhatdan) [inline Atomic/util.py]: Remove

File: Atomic/util.py [hunk 4]
```diff
@@ -267,10 +173,25 @@ def skopeo(image):
     else:
         return json.loads(results.stdout.decode('utf-8'))
 
+
 class NoDockerDaemon(Exception):
```
Comment (rhatdan) [inline Atomic/util.py]: Remove

File: Atomic/util.py [hunk 5]
```diff
@@ -267,10 +173,25 @@ def skopeo(image):
     else:
         return json.loads(results.stdout.decode('utf-8'))
 
+
```
Comment (rhatdan) [inline Atomic/util.py]: Remove

File: Atomic/util.py [hunk 6]
```diff
@@ -267,10 +173,25 @@ def skopeo(image):
...
+
+def get_atomic_config():
+    """
+    Returns the atomic configuration file (/etc/atomic.conf)
+    in a dict
+    :return: dict based structure of the atomic config file
+    """
+    atomic_conf = '/etc/atomic.conf'
+    if not os.path.exists(atomic_conf):
```
Comment (rhatdan) [inline Atomic/util.py]: Make /etc/atomic.conf a constant and then use it throughout the code.

File: atomic [hunk 7]
```diff
@@ -531,7 +533,9 @@ if __name__ == '__main__':
     except subprocess.CalledProcessError as e:
         sys.stderr.write("\n")
         sys.exit(e.returncode)
-    except AttributeError:
+    except AttributeError as e:
         # python3 throws exception on no args to atomic
```
Comment (rhatdan) [inline atomic]: Not using "e"?

File: atomic [hunk 8]
```diff
@@ -389,13 +390,14 @@ if __name__ == '__main__':
...
-    scanp.set_defaults(func='scan')
+    scanp.set_defaults(_class=Scan, func='scan')
     scan_group = scanp.add_mutually_exclusive_group()
     scan_out = scanp.add_mutually_exclusive_group()
     scan_out.add_argument("--json", default=False, action='store_true', help=_("output json"))
     scan_out.add_argument("--detail", default=False, action='store_true', help=_("output more detail"))
     scanp.add_argument("scan_targets", nargs='*', help=_("container image"))
     scanp.add_argument("--fetch_cves", type=bool_detect, default=None,  help=_("override the behavior defined in /etc/oscapd/config.ini as to whether to download the most recent CVE data. Values can be True of False"))
+    scanp.add_argument("--scanner", choices=['openscap', 'blackduck'], default='openscap', help=_("define the intended scanner"))
     scan_group.add_argument("--all", default=False, action='store_true', help=_("scan all images (excluding intermediate layers) and containers"))
```
Comment (rhatdan) [inline atomic]: These should be populated from /etc/atomic.conf not hard coded.

File: Atomic/util.py [hunk 9]
```diff
@@ -267,10 +173,25 @@ def skopeo(image):
     else:
         return json.loads(results.stdout.decode('utf-8'))
 
+
 class NoDockerDaemon(Exception):
     def __init__(self):
         Exception.__init__(self, "The docker daemon does not appear to be running.")
 
+
```
Comment (rhatdan) [inline Atomic/util.py]: remove

File: Atomic/atomic.py (hunk 2 above)
Comment (baude) [inline Atomic/atomic.py]: Done

File: Atomic/util.py (hunk 5 above)
Comment (baude) [inline Atomic/util.py]: pep8 conventions wants two blank lines between definitions not in a class.  still want me to remove.  this goes for several of the others you flagged.

File: Atomic/util.py (hunk 6 above)
Comment (baude) [inline Atomic/util.py]: Done, fixed.

File: atomic (hunk 7 above)
Comment (baude) [inline atomic]: Fixed, nice catch.

File: Atomic/util.py (hunk 5 above)
Comment (rhatdan) [inline Atomic/util.py]: No I guess if that is the convention, we should go with it.

File: Atomic/atomic.py (hunk 1 above)
Comment (baude) [inline Atomic/atomic.py]: Can you elaborate on this one.  I dont quite understand.

Comment (baude) [inline Atomic/atomic.py]: Done, fixed

File: atomic (hunk 8 above)
Comment (baude) [inline atomic]: Done, fixed

**Comment by sghosh151** [comment]
Is there a way to set the configuration via cloud-init?

**Comment by rhatdan** [comment]
@sghosh151 Please explain further what you want?

**Comment by baude** [comment]
bot, retest this please

**Comment by baude** [comment]
Ok added review comments from Dan

File: Atomic/atomic.py [hunk 10]
```diff
@@ -843,31 +749,39 @@ def get_image_infos(self):
...
+                                       bind=True)
 
         except (mount.MountError, mount.NoDockerDaemon) as dme:
             raise ValueError(str(dme))
 
-    def unmount(self):
+    def unmount(self, mountpoint=None):
+        if mountpoint is None:
+            mountpoint = self.args.mountpoint
         try:
```
Comment (rhatdan) [inline Atomic/atomic.py]: Can we just do
```
 def unmount(self, mountpoint=self.args.mountpoint):
```
Is that legal?

File: Atomic/util.py [hunk 11]
```diff
@@ -6,10 +6,18 @@
...
 """Atomic Utility Module"""
 
 ReturnTuple = collections.namedtuple('ReturnTuple',
                                      ['return_code', 'stdout', 'stderr'])
+custom_conf = os.environ.get('ATOMIC_CONF')
+
+if custom_conf is None:
+    ATOMIC_CONF = '/etc/atomic.conf'
+else:
+    ATOMIC_CONF = custom_conf
```
Comment (rhatdan) [inline Atomic/util.py]: We have decided to make this use /etc/atomic.d  So people can drop multiple config files into this directory.

File: atomic [hunk 12]
```diff
@@ -389,13 +392,13 @@ if __name__ == '__main__':
...
         epilog="atomic scan <input> scans a container or image for CVEs")
-    scanp.set_defaults(func='scan')
+    scanp.set_defaults(_class=Scan, func='scan')
     scan_group = scanp.add_mutually_exclusive_group()
     scan_out = scanp.add_mutually_exclusive_group()
-    scan_out.add_argument("--json", default=False, action='store_true', help=_("output json"))
-    scan_out.add_argument("--detail", default=False, action='store_true', help=_("output more detail"))
     scanp.add_argument("scan_targets", nargs='*', help=_("container image"))
     scanp.add_argument("--fetch_cves", type=bool_detect, default=None,  help=_("override the behavior defined in /etc/oscapd/config.ini as to whether to download the most recent CVE data. Values can be True of False"))
+    scanp.add_argument("--scanner", choices=[x for x in atomic_config['scanners']], default='openscap', help=_("define the intended scanner"))
```
Comment (rhatdan) [inline atomic]: Not sure we should have --fetch_cves any longer, this is specific to one container type.  Should be inside the container.

**Comment by rhatdan** [comment]
We need updates to bash command line completions also.

File: Atomic/atomic.py (hunk 10 above)
Comment (rhatdan) [inline Atomic/atomic.py]: Nope it is not, ignore

File: Makefile [hunk 13]
```diff
@@ -59,5 +59,9 @@ install-only:
 
 	echo ".so man1/atomic-push.1" > $(PREFIX)/share/man/man1/atomic-upload.1
 
+	install -m 644 atomic.conf $(DESTDIR)/etc
+
+	install -d $(DESTDIR)/atomic.d
```
Comment (jlebon) [inline Makefile]: Shouldn't this be $(DESTDIR)/etc/atomic.d ?

File: Atomic/util.py [hunk 14]
```diff
@@ -7,10 +7,24 @@
...
+
+if custom_confd is None:
+    ATOMIC_CONFD = '/etc/atomic.d/'
+else:
+    ATOMIC_CONFD = custom_confd
```
Comment (jlebon) [inline Atomic/util.py]: You can simplify this by passing the fallback var to os.environ.get()

File: docs/atomic.1.md [hunk 15]
```diff
@@ -15,6 +15,10 @@ Atomic Management Tool
 # OPTIONS
 **-h** **--help**
   Print usage statement
+# ENVIRONMENT VARIABLES
+
+**ATOMIC_CONF** The location of the atomic configuration file (normally /etc/atomic.conf) can be
+overriden with the _ATOMIC_CONF_ environment variable
```
Comment (jlebon) [inline docs/atomic.1.md]: Missing an entry for ATOMIC_CONFD here. Though I think a cleaner approach might be to just have a single ATOMIC_ETC here and construct the atomic.conf and atomic.d paths from that.

File: Atomic/util.py (hunk 14 above)
Comment (jlebon) [inline Atomic/util.py]: E.g. `os.environ.get('ATOMIC_CONF', "/etc/atomic.conf")`

File: Atomic/util.py [hunk 16]
```diff
@@ -262,16 +180,48 @@ def skopeo(image):
...
         Exception.__init__(self, "Unable to associate '{}' with an image or container".format(msg))
+
+
+def get_atomic_config():
+    """
+    Returns the atomic configuration file (/etc/atomic.conf)
+    in a dict
+    :return: dict based structure of the atomic config file
+    """
+    if not os.path.exists(ATOMIC_CONF):
```
Comment (jlebon) [inline Atomic/util.py]: custom_conf?

File: Atomic/util.py [hunk 17]
```diff
@@ -262,16 +180,48 @@ def skopeo(image):
...
+        raise ValueError("{} does not exist".format(ATOMIC_CONF))
+    with open(ATOMIC_CONF, 'r') as conf_file:
+        return yaml_load(conf_file)
+
+def get_scanners():
+    scanners = []
+    if not os.path.exists(ATOMIC_CONFD):
```
Comment (jlebon) [inline Atomic/util.py]: custom_confd?

File: Atomic/util.py (hunk 14 above)
Comment (baude) [inline Atomic/util.py]: Great comment, fixed

File: Makefile (hunk 13 above)
Comment (baude) [inline Makefile]: Great catch, fixed.

File: Atomic/util.py (hunk 16 above)
Comment (baude) [inline Atomic/util.py]: Reworked.

File: Atomic/util.py (hunk 17 above)
Comment (baude) [inline Atomic/util.py]: Reworked.

File: docs/atomic.1.md (hunk 15 above)
Comment (baude) [inline docs/atomic.1.md]: It's coming on the next push

**Comment by baude** [comment]
@rhatdan  This is now the reworked PR.  Passes tests.  python3 wfm.

**Comment by rhatdan** [comment]
LGTM

**Comment by rhatdan** [comment]
@cgwalters-bot r+ 6ed4994

**Comment by rhatdan** [comment]
@baude We need man pages for atomic.conf and atomic.conf.d

**Pull request merged**